---
title: "Azolla genus wide metagenome assembly taxonomy and statistics"
author: "Laura Dijkhuizen"
date: "03/12/2021"
output:
  pdf_document: default
  html_notebook: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/")
```


 
# Aim
In this document I aim to explore and discribe several metagenome assembly statistics for several metagenome assemblies derived from sequencing data of Azolla ferns. 
The statistics I work with are contig/scaffold length, depth (abundance), taxonomy, and ORF count. 
Later I'll add ORF length distributions as well, but that data needs to be generated first.
The use of discribing metagenome assemblies as whole assemblies, is to compare these assemblies agnostic of the binning process (and choices made in that process).
Based on just a best estimate of taxonomy and depth, similarities may be found and species diversity may be explored comparing all host species simultaneously.
Secondly, I aim to evaluate whether the extra filtering, then reassembly step improved the assembly.

## TODO

* make collapsible html page
* make pdf output
* ORF distributions per assembly
* ORF distributions per assembly, per kingdom
* rearange document to simultaneously process single and hybrid libraries
* discribe packages used
* CAT reference
* plot kmer depth vs backmap depth


# import data 
Contig/scaffold length and depth is taken from the headers of the Spades metagenome assembly output. 
This may not be as 'nice' as mapping the original reads to the assembly file, but it is substantially quicker and in my experience it works very well.
During import, I discard any contigs that are shorter than 2500bp for I consider them too short for binning.
Binning software usually discards short contigs anyway, 2500bp is the threshold used by metabat2.

The original data is stored in separate files for single library assemblies and hybrid assemblies. Here I import both:

```{r import libraries and data}
#setwd
setwd("/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/")
library(data.table)
library(ggplot2)
header <-  c('assembly', 'sample', 'scaffolded', 'node',    'length',  'coverage', 'ORFs',    'ORFs_classified', 'superkingdom', 'phylum', 'class',  'order',  'family', 'genus', 'species')
classes <- c('factor',   'factor', 'factor',     'numeric', 'numeric', 'numeric',  'numeric', 'numeric',         'factor',       'factor', 'factor', 'factor', 'factor', 'factor', 'factor')
metrics_single <- fread(input = "/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/assembly_stats_and_taxonomy.tab",
                 sep = '\t',
                 fill = TRUE,
                 header = F,
                 stringsAsFactors = T,
                 data.table = T,
                 col.names = header,
                 colClasses = classes
)[length >= 2500]
```
And now the hybrid file:
```{r import libraries and data hybrid}
metrics_hybrid <- fread(input = "/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/assembly-hybrid_stats_and_taxonomy.tab",
                 sep = '\t',
                 fill = TRUE,
                 header = F,
                 stringsAsFactors = T,
                 data.table = T,
                 col.names = header,
                 colClasses = classes
)[length >= 2500]
```

## Data export
Next, for convenience, I'd like to merge the two datasets, and write them back to a smaller file to use for a shiny app later on.
```{r}
fwrite(x = metrics_single,file = "/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/assembly_stats_and_taxonomy_2500.tab",sep="\t",append = F,quote = F,col.names = F)
fwrite(x = metrics_hybrid,file = "/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/assembly_stats_and_taxonomy_2500.tab",    sep="\t",append = T,quote = F,col.names = F)
```
And afterwards, I test if the data imports cleanly:
```{r}
metrics <- fread(input = "/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/assembly_stats_and_taxonomy_2500.tab", 
                 sep = '\t',
                 fill = TRUE,
                 header = F,
                 stringsAsFactors = T,
                 data.table = T,
                 col.names = header,
                 colClasses = classes
)
```

For testing purposes, this is how the data is cleaned in the shiny app:
```{r}
singlelibs <- c("Azfil_lab_250", "Azfil_lab_500", "Azfil_lab_800", "Azfil_minuscyano_170", "Azfil_minuscyano_350", "Azfil_wild_galgw_E_1", "Azfil_wild_galgw_E_2", "Azfil_wild_galgw_E_3", "Azfil_wild_galgw_P_2", "Azfil_wild_galgw_P_3", "Azfil_wild_galgw_P_4") 
hybridlibs <- c('Azfil_wild', 'Azfil_minuscyano', 'Azfil_lab')
metrics$assemblytype <- factor(x = 'single library',levels = c('single library','hybrid library','partial library'))
metrics[sample %in% singlelibs]$assemblytype <- 'partial library'
metrics[sample %in% hybridlibs]$assemblytype <- 'hybrid library'
```

cleanup:
```{r}
rm(classes,header)
rm(singlelibs,hybridlibs)
```

# Assembly samples
In this stats and taxonomy file there is information about `r length(levels(metrics$assembly))` types of assemblies being: `r levels(metrics$assembly)`
Irrespective these assembly types there are `r length(levels(metrics$sample))` samples or 'hostcodes' als they are called in the Snakemake workflow.
The single libraries table counts `r length(levels(metrics_single$sample))` samples in assembly types: `r levels(metrics_single$assembly)`.

# Assembly size

## Assembly size
Overall, looking at the scaffolded assemblies, I've assembled nearly 920mbase of DNA with a median size of 4.2kb.
The size distribution is clearly skewed towards the short end of the spectrum, for the mean size is much bigger.
```{r}
knitr::kable(
  metrics[assembly == 'singles_doublefiltered' | assembly == 'hybrid_doublefiltered' , 
          .(contig_count=length(length), 
            length_mb=sum(length)/1000000, 
            median_length_kb=median(length),
            average_size_kb=mean(length)
            ), 
          .(assembly,scaffolded)]
  ,digits = 0
)
```
and the same numbers per sample:
```{r}
knitr::kable(
  metrics[assembly == 'singles_doublefiltered' & scaffolded == 'scaffolds', 
          .(contig_count=length(length), 
            length_mb=sum(length)/1000000, 
            median_length_kb=median(length),
            average_size_kb=mean(length)
            ), 
          sample]
  ,digits = 0
)
```

## length distributions
Let's plot the length distribution for the scaffolds and contigs assembled in the different assemblies.
Contigs or scaffolds shorter than 2.5kbp are discarded by default in metabat2 and by default not loaded in this R script.
```{r lengtt hplot_linear, fig.height = 8, fig.width = 15}
lengthplot<- ggplot (metrics,aes(x=length,col=scaffolded,alpha=.4))
lengthplot <- lengthplot + geom_freqpoly(bins=50)
lengthplot <- lengthplot + facet_grid(assembly ~ sample)
#lengthplot <- lengthplot + scale_y_continuous(trans = 'log10')
lengthplot <- lengthplot + scale_x_continuous(trans = 'log10')
lengthplot <- lengthplot + theme_classic()
lengthplot <- lengthplot + labs(y='count',x='sequence length')
lengthplot <- lengthplot + theme(legend.position = "bottom",
                                strip.text.x = element_text(angle = 80),
                                axis.text.x = element_text(angle = 80,hjust = 1))
lengthplot
```

The difference between length distributions of scaffolds and contigs is minimal and distributed as expected; scaffold frequency lines peak a bit to the right of contig frequency lines.
Plotting the lengths as boxplots however shows the subtle differences a bit better: the upper quartile often is a bit higher for scaffolded assemblies, or when it is lower, the scaffolded assemblies of more outliers in the high end of the boxplots.
The lower end of the length distribution does not benefit from scaffolding as much, but also to such an extent that the scaffolded assemblies contain more scaffolds than the contig assemblies contain contigs.
We'll learn later that the minuscyano samples are very lowly diverse and sufficiently saturated to cover all microbes with substantial depth.
This is my explanaition for the substantial higher length distribution for minuscyano single library assemblies.
The hybrid assemblies show less of a difference amongst each other, supporting the theory that the short contigs are caused by graph breaks of lowly abundant genomes.

```{r lengthboxplot_2500_log10, fig.height = 8, fig.width = 15}
#names(metrics2500)
lenghtbox <- ggplot(metrics,aes(y=length,x=sample,col=scaffolded,alpha=.4))
lenghtbox <- lenghtbox + geom_boxplot()
#lenghtbox <- lenghtbox + geom_violin(draw_quantiles = T)
lenghtbox <- lenghtbox + facet_grid(assembly ~ . )
lenghtbox <- lenghtbox + theme_classic()
lenghtbox <- lenghtbox + scale_y_log10()
lenghtbox <- lenghtbox + theme(axis.text.x = element_text(angle = 45,hjust = 1))
lenghtbox
```


# Assembly filtering
I wonder if scaffolds after the double filter are longer than if they are only filtered for host DNA. 
First I plot this as a table, and later the length distributions in a graph.

```{r}
knitr::kable(
  metrics[scaffolded == 'scaffolds' & assembly != 'hybrid_doublefiltered',
          .(contig_count=length(length), 
            length_mb=sum(length)/1000000, 
            median_length_kb=median(length),
            average_size_kb=mean(length)
            ), 
          .(superkingdom,assembly)]
  ,digits = 0
)
```
Based on this table, summing assembly sizes for all samples, the filtering had a negligible impact on bacterial assembly size, but drastically reduced eukaryotic assembly size.
The not classified and empty category were also reduced in size, but not as drastic as the eukaryotic set.
This makes perfect sense for I used that taxonomy indication to filter out reads again, what's surprising is that still eukaryotic contigs remain.
Additionally, viruses and archaea are also reduced in assembly size. 
This is not a wanted effect, so for these kingdoms specifically (if classified correctly) I'll refer to the original rather than the double filtered assemblies when that time comes.
Based on these tables, assembly length and median/mean lenghts have not improved.

Now let's plot this as boxplots.
Here, I'm splitting the contigs based on superkingdom, and for legibility, I'm not displaying contigs which were not classified.
I do include archaea and viruses, even though their numbers are limited.
Naturally the distributions will differ for sequences identified by CAT as plant were filtered out. Hence, I plot these separately.
```{r lengthboxplot_2500+_perkingdom, fig.height = 10, fig.width = 15}
lenghtbox <- ggplot(metrics[scaffolded == 'scaffolds' & 
                              assembly != 'hybrid_doublefiltered' & 
                              superkingdom != '' & 
                              superkingdom != 'not classified'],
                    aes(y=length/1000,col=assembly,alpha=.4)
                    )
lenghtbox <- lenghtbox + facet_grid(superkingdom ~ sample,scales = 'free_y',space = 'free_y',)
#lenghtbox <- lenghtbox + geom_violin()
lenghtbox <- lenghtbox + geom_boxplot()
#lenghtbox <- lenghtbox + geom_freqpoly(inherit.aes = F, aes(x=length,col=assembly,alpha=.4),binsize=80 ) 
lenghtbox <- lenghtbox + theme_classic()
lenghtbox <- lenghtbox + scale_y_log10()
#lenghtbox <- lenghtbox + geom_vline(xintercept =0)
lenghtbox <- lenghtbox + theme(axis.text.x=element_text(angle = 90,hjust = 1),legend.position = "bottom")
lenghtbox
```

To my surprise, filtering out plant DNA from the first assembly, filtering the reads and then reassembling the plant DNA depleted short sequences, has not yielded longer scaffolds nor contigs for Bacteria. 
In hindsight, it became apparent to me that the assembly graphs of these drastically different taxa would not overlap in the assembly process, hence not ambiguity (in bacterial contigs) is filtered in this process, therefore no longer bacterial contigs are spit out of the assembly.
On the bright side of things, the double filter did effectively reduce the eukaryotic fraction of the assembly whilst not impacting the bacterial fraction.
There might be no direct benefit, other than making a hybrid assembly feasible where that was not the case with unfiltered nor host_filtered samples due to RAM constraints.

# Taxonomy

```{r tax_dotplot_kingdom, fig.height = 10, fig.width = 15 }
length_dist <- ggplot(metrics2500[scaffolded == 'scaffolds'],aes(x=length,y=coverage,size=length,alpha=.001,col=superkingdom))
length_dist <- length_dist + facet_grid(assembly ~ sample,scales = 'free_x')
length_dist <- length_dist + geom_point()
length_dist <- length_dist + scale_x_log10()
length_dist <- length_dist + scale_y_log10()
length_dist <- length_dist + labs(x='contig length',y='contig coverage')
length_dist <- length_dist + theme_classic()
length_dist <- length_dist + theme(legend.position = "bottom",
                                   strip.text.x = element_text(angle = 80),
                                   axis.text.x = element_text(angle = 80,hjust = 1))
length_dist
```
And save to a png file
```{r tax_dotplot_kingdom_png}
ggsave(filename = "./coverage_overlength_distributions_per_assembly_kingdom.png",
       plot = length_dist,
       device = 'png',
       dpi = 500,
       units = 'cm',
       width = 42,
       height = 29.7 )
```

Take out all eukaryota and plot the remaing phyla

```{r tax_dotplot_phylum, fig.height = 15, fig.width = 15 }
length_dist <- ggplot(metrics2500[scaffolded == 'scaffolds' & superkingdom != 'Eukaryota'],aes(x=length,y=coverage,size=length,alpha=.001,col=phylum))
length_dist <- length_dist + facet_grid(assembly ~ sample,scales = 'free_x')
length_dist <- length_dist + geom_point()
length_dist <- length_dist + scale_x_log10()
length_dist <- length_dist + scale_y_log10()
length_dist <- length_dist + labs(x='contig length',y='contig coverage')
length_dist <- length_dist + theme_classic()
length_dist <- length_dist + theme(legend.position = "bottom",
                                   strip.text.x = element_text(angle = 80),
                                   axis.text.x = element_text(angle = 80,hjust = 1))
length_dist
```

And save to a png file
```{r}
ggsave(filename = "./coverage_overlength_distributions_per_assembly_phylum.png",
       plot = length_dist,
       device = 'png',
       dpi = 500,
       units = 'cm',
       width = 42,
       height = 29.7 )
```


```{r tax_dotplot_order, fig.height = 10, fig.width = 15 }
metrics_temp <- metrics2500[scaffolded == 'scaffolds' & superkingdom != 'Eukaryota' & assembly == 'singles_doublefiltered']
droplist <- metrics_temp[, .(uniques=length(genus)) < 10, by=order]$order[ metrics_temp[, .(uniques=length(genus)) < 50, by=order]$V1 == T ]
droplist <- as.character(droplist)
metrics_temp[order == "Chromatiales" | order == "Salinisphaerales" | order == "Neisseriales" | order == "Enterobacterales"  | order == "Bacteroidales" | order ==  "Solirubrobacterales"  | order ==  "Vibrionales" | order ==   "Sneathiellales"   | order ==       "Methylococcales"   | order == "Candidatus Babeliales*" | order == "Streptomycetales"  | order == "Nitrospirales"   | order ==   "Corynebacteriales"  | order ==  "Chloroflexales"  | order ==    "Selenomonadales"  | order ==   "Lactobacillales"  | order ==   "Alteromonadales"  | order == "Oceanospirillales"   | order == "Cellvibrionales"  | order ==   "Geodermatophilales"  | order ==  "Ortervirales"     | order ==   "Aeromonadales"  | order ==     "Myxococcales"  | order == "Chroococcales"   | order ==    "Entomoplasmatales"   | order ==   "Chroococcidiopsidales" | order ==  "Solibacterales"  | order == "Planctomycetales"  | order == "Desulfovibrionales" | order == "Pseudonocardiales" | order ==  "Verrucomicrobiales"   | order ==  "Chlorobiales"  | order ==    "Clostridiales"  | order ==  "Marinilabiliales"  | order ==  "Holosporales"   | order == "Desulfuromonadales" | order ==   "Silvanigrellales"  | order ==   "Acidiferrobacterales"  | order == "Capsulimonadales*"   | order == "Pseudomonadales" | order == "Xanthomonadales" | order ==  "Rhodobacterales" | order =="Flavobacteriales"  | order =="Propionibacteriales" | order == "Synechococcales" | order == "Frankiales", order:= c('other')]

library(plotly)

length_dist <- ggplot(metrics_temp,aes(x=length,y=coverage,size=length,alpha=.001,col=order))
length_dist <- length_dist + facet_grid( ~ sample,scales = 'free_x')
length_dist <- length_dist + geom_point()
length_dist <- length_dist + scale_x_log10()
length_dist <- length_dist + scale_y_log10()
length_dist <- length_dist + labs(x='contig length',y='contig coverage')
length_dist <- length_dist + theme_classic()
length_dist <- length_dist + guides(col=guide_legend(ncol=5))
length_dist <- length_dist + theme(legend.position = "bottom",
                                   strip.text.x = element_text(angle = 80),
                                   axis.text.x = element_text(angle = 80,hjust = 1))
length_dist
```


```{r}
ggplotly(p=length_dist)
```


```{r}
metrics_temp[, .(uniques=length(genus)) , by=order]
rm(metrics_temp)
```



```{r}
ggsave(filename = "./coverage_overlength_distributions_per_assembly_order.png",
       plot = length_dist,
       device = 'png',
       dpi = 500,
       units = 'cm',
       width = 21/9*16,
       height = 21 )
```




# Assembly samples
In this stats and taxonomy file there is information about `r length(levels(metrics$assembly))` types of assemblies being: `r levels(metrics$assembly)`
Irespective these assembly types there are `r length(levels(metrics$sample))` samples or 'hostcodes' as they are called in the Snakemake workflow.

# Assembly length
Let's plot the length distribution for the scaffolds and contigs assembled in the different assemblies. Contigs or scaffolds shorter than 2.5kbp are discarded by default in metabat2, the red vertical line indicates this separation.
```{r lengthplot_linear_hybrid, fig.height = 5, fig.width = 10}
lengthplot <- ggplot (metrics_hybrid,aes(x=length,col=scaffolded,alpha=.4))
lengthplot <- lengthplot + geom_freqpoly(bins=50)
lengthplot <- lengthplot + facet_grid(assembly ~ sample)
lengthplot <- lengthplot + scale_y_continuous(trans = 'log10')
lengthplot <- lengthplot + scale_x_continuous(trans = 'log10')
lengthplot <- lengthplot + theme_classic()
lengthplot <- lengthplot + labs(y='count',x='sequence length')
lengthplot <- lengthplot + geom_vline(xintercept = 2500,colour='red')
lengthplot <- lengthplot + theme(legend.position = "bottom",
                                 strip.text.x = element_text(),
                                 axis.text.x = element_text(angle = 80,hjust = 1))
lengthplot
```

the double log10 axis slightly exagerates the amount of sequencing lost, ploted in a linear x-axis it is more fair to judge.
```{r lengthplot_log10_hybrid, fig.height = 5, fig.width = 10}
lengthplot <- ggplot(metrics_hybrid,aes(x=length,col=scaffolded,alpha=.4))
lengthplot <- lengthplot + geom_freqpoly(bins=50)
lengthplot <- lengthplot + facet_grid(assembly ~ sample)
lengthplot <- lengthplot + scale_y_continuous(trans = 'log10')
#lengthplot <- lengthplot + scale_x_continuous(trans = 'log10')
lengthplot <- lengthplot + theme_classic()
lengthplot <- lengthplot + labs(y='count',x='sequence length')
lengthplot <- lengthplot + geom_vline(xintercept = 2500,colour='red')
lengthplot <- lengthplot + theme(legend.position = "bottom",
                                 strip.text.x = element_text(angle = 80),
                                 axis.text.x = element_text(angle = 80,hjust = 1))
lengthplot
```


Create a 2500+ subset.

```{r subset to 2500 hybrid}
metrics2500_hybrid <- metrics_hybrid[length >= 2500]
```

And plot:

```{r lengthboxplot_2500_log10_hybrid, fig.height = 5, fig.width = 8}
lenghtbox <- ggplot(metrics2500_hybrid,aes(y=length,x=sample,col=scaffolded,alpha=.4))
lenghtbox <- lenghtbox + geom_boxplot()
lenghtbox <- lenghtbox + facet_grid(assembly ~ . )
lenghtbox <- lenghtbox + theme_classic()
lenghtbox <- lenghtbox + scale_y_log10()
lenghtbox <- lenghtbox + theme(axis.text.x = element_text(angle = 45,hjust = 1))
lenghtbox
```

```{r lengthboxplot_2500+_perkingdom_hybrid, fig.height = 10, fig.width = 10}
lenghtbox <- ggplot(metrics2500_hybrid,
                    aes(y=length,x=sample,col=sample,alpha=.4)
                    )
lenghtbox <- lenghtbox + facet_grid(scaffolded ~ superkingdom)
#lenghtbox <- lenghtbox + geom_violin()
lenghtbox <- lenghtbox + geom_boxplot()

lenghtbox <- lenghtbox + theme_classic()
lenghtbox <- lenghtbox + scale_y_log10()
lenghtbox <- lenghtbox + geom_vline(xintercept =0)
lenghtbox <- lenghtbox + theme(axis.text.x=element_text(angle = 90,hjust = 1),legend.position = "bottom")
lenghtbox
```















# Taxonomy

```{r tax_dotplot_kingdom_hybrid, fig.height = 8, fig.width = 8 }
length_dist <- ggplot(metrics2500_hybrid[scaffolded == 'scaffolds'],aes(x=length,y=coverage,size=length,alpha=.001,col=superkingdom))
length_dist <- length_dist + facet_grid(assembly ~ sample,scales = 'free_x')
length_dist <- length_dist + geom_point()
length_dist <- length_dist + scale_x_log10()
length_dist <- length_dist + scale_y_log10()
length_dist <- length_dist + labs(x='contig length',y='contig coverage')
length_dist <- length_dist + theme_classic()
length_dist <- length_dist + theme(legend.position = "bottom",
                                   strip.text.x = element_text(angle = 80),
                                   axis.text.x = element_text(angle = 80,hjust = 1))
length_dist
```
And save to a png file
```{r tax_dotplot_kingdom_png_hybrid}
ggsave(filename = "./coverage_hybrid_overlength_distributions_per_assembly_kingdom.png",
       plot = length_dist,
       device = 'png',
       dpi = 500,
       units = 'cm',
       width = 42,
       height = 29.7 )
```

Take out all eukaryota and plot the remaing phyla

```{r tax_dotplot_phylum_hybrid, fig.height = 10, fig.width = 8 }
length_dist <- ggplot(metrics2500_hybrid[scaffolded == 'scaffolds' & superkingdom != 'Eukaryota'],aes(x=length,y=coverage,size=length,alpha=.001,col=phylum))
length_dist <- length_dist + facet_grid(assembly ~ sample,scales = 'free_x')
length_dist <- length_dist + geom_point()
length_dist <- length_dist + scale_x_log10()
length_dist <- length_dist + scale_y_log10()
length_dist <- length_dist + labs(x='contig length',y='contig coverage')
length_dist <- length_dist + theme_classic()
length_dist <- length_dist + theme(legend.position = "bottom",
                                   strip.text.x = element_text(angle = 80),
                                   axis.text.x = element_text(angle = 80,hjust = 1))
length_dist
```

And save to a png file
```{r}
ggsave(filename = "./coverage_hybrid_overlength_distributions_per_assembly_phylum.png",
       plot = length_dist,
       device = 'png',
       dpi = 500,
       units = 'cm',
       width = 42,
       height = 29.7 )
```


```{r tax_dotplot_order_hybrid, fig.height = 10, fig.width = 10 }
length_dist <- ggplot(metrics2500_hybrid[scaffolded == 'scaffolds' & superkingdom != 'Eukaryota'],aes(x=length,y=coverage,size=length,alpha=.001,col=order))
length_dist <- length_dist + facet_grid(. ~ sample,scales = 'free_x')
length_dist <- length_dist + geom_point()
length_dist <- length_dist + scale_x_log10()
length_dist <- length_dist + scale_y_log10()
length_dist <- length_dist + labs(x='contig length',y='contig coverage')
length_dist <- length_dist + theme_classic()
length_dist <- length_dist + guides(col=guide_legend(ncol=5))
length_dist <- length_dist + theme(legend.position = "bottom")
length_dist
```


```{r}
ggsave(filename = "./coverage_hybrid_overlength_distributions_per_assembly_order.png",
       plot = length_dist,
       device = 'png',
       dpi = 500,
       units = 'cm',
       width = 42,
       height = 29.7 )
```


# packages
* datatable
* ggplot2
* kable

# Bash script

