---
title: "Statistics on Bins in the Azolla metagenome project"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---


# Import data for 'Refined' collection of bins/MAGs.
Import the data, firstly a table of CheckM, BAT and Anvi'o data combined in snakemake. 
For now I only import the refined dataset, since this is the main piece of work, but later I'll work on comparing the different binning options.
```{r, root.dir="/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/"}
#setwd("/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/")
library(data.table)
library(ggplot2)
bins <- fread(input = "/stor/azolla_metagenome/Azolla_genus_metagenome/data/curated_bins/refined.bin_info.tab",
              sep = '\t',
              fill = TRUE,
              header = T,
              stringsAsFactors = T,
              data.table = T
)
```

## Azfil wild sample
Additionally, I'm interested in calculing the differential abundance for all bins in the _Azolla filiculoides_ 'wild' sample.
Since this is the only sample with differential sampling, I'll do this separately from other analyses.
```{r}
azfil_wild_abundance <- fread(input = "../data/curated_bins/refined/Azfil_wild/bins_across_samples/mean_coverage.txt",
                              sep = '\t',
                              header= T,
                              stringsAsFactors = T,
                              data.table = T,
                              fill= T
                              )
names(azfil_wild_abundance) <- sub('assembly_hybrid_doublefiltered_sample_Azfil_wild_binningsignal_',
                                   '',
                                   names(azfil_wild_abundance)
                                   )

Azfil_ratios <- azfil_wild_abundance[,.(bin=bins,
                                        ratio_leaf_over_plant=((Azfil_wild_galgw_E_1+Azfil_wild_galgw_E_2+Azfil_wild_galgw_E_3)/3)
                                                             /((Azfil_wild_galgw_P_2+Azfil_wild_galgw_P_3+Azfil_wild_galgw_P_4)/3),
                                        pacbio_mean_coverage=pacbio_reads)]
Azfil_ratios[,.(bin,
                ratio_leaf_over_plant=round(ratio_leaf_over_plant,2),
                pacbio_mean_coverage=round(pacbio_mean_coverage,2))]
```
That's that, I include here presence ratio of leaf samples over plant samples, as well as PacBio read presence.
The rationalle behind the latter that this sequencing was derived from plants sampled at the same spot but a few years earlier.
Hence, for a bin to be present in both samples, and be enriched in the leaf samples, is a strong indication that it is endophytic.

```{r}
bins[(Sample == 'Azfil_wild'),
     .(order,
       length_mb     = round(sum(total_length/1000000),2),
       anvio_compl   = round(sum(percent_completion),0),
       anvio_redun   = round(sum(percent_redundancy),0),
       checkm_compl  = round(sum(Completeness),0),
       checkm_redun  = round(sum(Contamination),0)
       ),
     by=c('Sample', 'bins')
     ]
```
Above is a table with all kinds of stats for the Azfil wild sample, this is going to be a maintext table so I write the code separately here.

## Bins, size, taxonomy and QC for the entire genus
Now, look at a table with all bins, their origin sample, size, completion and redundancy.
~~I do this for everything except Azfil wild, since I did that already above.~~

This table is huge, so I don't include it in the markdown render here, instead I write it to csv so I can make something out of it.
```{r}
write.csv(
  x = bins[,#(Sample != 'Azfil_wild'),
     .(length_mb     = round(sum(total_length/1000000),2),
       anvio_compl   = round(sum(percent_completion),0),
       anvio_redun   = round(sum(percent_redundancy),0),
       checkm_compl  = round(sum(Completeness),0),
       checkm_redun  = round(sum(Contamination),0)
       ),
     by=c('Sample', 'bins')
     ],file = "all_bins_table.csv",
  row.names = T)
```

```{r}
bins[,
     .(order=order,
       length_mb     = round(sum(total_length/1000000),2),
       anvio_compl   = round(sum(percent_completion),0),
       anvio_redun   = round(sum(percent_redundancy),0),
       checkm_compl  = round(sum(Completeness),0),
       checkm_redun  = round(sum(Contamination),0)
       ),
     by=c('Sample', 'bins')
     ]
```


## Look at bins per sample
Now let's see how many bins we have per sample.
I'll make separate tables for various QC cut-offs and combine these in Markdown/LateX.
```{r}
bins_per_sample <- bins[,
                        .(bincount=length(bins)),
                        by='Sample'
                        ]
bins_per_sample[order(-bins_per_sample[,2])]
```

How many bins do we have per sample with sufficient quality according to anvio
```{r}
bins[(percent_completion > 70 & percent_redundancy < 20),
                        .(bincount=length(bins)),
                        by='Sample'
                        ][order(-bins_per_sample[,2])]
```

How many bins do we have per sample with sufficient quality according to CheckM
```{r}
bins[(Completeness > 70 & Contamination < 20),
                        .(bincount=length(bins)),
                        by='Sample'
                        ][order(-bins_per_sample[,2])]
```

## bin vs metagenome size
Now I wonder how much of a total metagenome is present in bins. 
First I need to get the total size of a metagenome, this is most easily done from the fasta headers of the spades assemblies. 
While doing so, I want to filter out any contigs shorter dan 2.5kbp, since these aren't binned anyway.
I do this in bash:

```{bash}
cd /stor/azolla_metagenome/Azolla_genus_metagenome
fastas=( data/assembly_*doublefiltered/*/scaffolds.fasta )
for f in ${fastas[@]}
do  echo $f':' $(                    \
    cat $f                           \
    | grep '>'                       \
    | egrep -o 'length_[0-9]+'       \
    | sed 's/length_//g'             \
    | awk '$1 > 2500  {print $1 ;}'  \
    | awk '{s+=$1} END {print s/1000000}' )
done
```
Now I put these manually in a df, not very elegant but I prefer it over having another small random tab file. 
Next, I'll calculate fraction binned with that data.
```{r}
fraction_binned <- bins[,
                        .(binned_mb     = round(sum(total_length/1000000),2)
                          ),
                        by=c('Sample')
                        ]
fraction_binned$assembly_mb <- c(56.9081,
                                 115.874,
                                 36.5311,
                                 179.447,
                                 104.468,
                                 100.126,
                                  95.626,
                                  90.133,
                                  79.409
                                )
fraction_binned$fraction <- round(fraction_binned$binned_mb / fraction_binned$assembly_mb * 100)
fraction_binned
```
Actually, I don't believe these numbers, I'm taking them from the Anvio summary html pages instead.

## Bin taxonomy
Next, pivot sample origin versus taxonomy and either count bins, or display size in mbase

First, count bins in class:
```{r}
dcast.data.table(data = bins[class != NA | class != 'no_support']
                 ,formula = Sample ~ class 
                 ,fun.aggregate = length
                 ,value.var = 'total_length'
                 )
```

Second, count bins in order and show only orders that have more than 5 bins over the entire genus metagenome.
```{r,}
#knitr::kable(
tmpdt <- dcast.data.table(data = bins[order != NA | order != 'no_support']
                          ,formula = order ~ Sample
                          ,fun.aggregate = length
                          ,value.var = 'total_length'
                          )
#data.table::all
#knitr::kable(tmpdt)
threshold <- 5
vector <- (apply(X = tmpdt[,2:length(tmpdt)],MARGIN = 1,FUN = 'sum') >= threshold)
tmpdt[vector]
```

Now select the bins from azfil lab that meet the above threshold (and remove the nostocales for now)
```{r}
#interesting orders:
i_orders <- tmpdt[(vector & order != 'Nostocales')]$order

bins[Sample == 'Azfil_lab' &
     order  %in% i_orders  ]$bins
```

Now let's look at that table with all bins again, but filter for the orders we defined above

```{r}
bins[order %in% i_orders,
     .(length_mb     = round(sum(total_length/1000000),2),
       anvio_compl   = round(sum(percent_completion),0),
       anvio_redun   = round(sum(percent_redundancy),0),
       checkm_compl  = round(sum(Completeness),0),
       checkm_redun  = round(sum(Contamination),0)
       ),
     by=c('Sample', 'bins')
     ]
```

# binning quality per sample per method figure
As a final figure for the chapter, I want to visualise somehow if the quality of binning improved with my manual selection.
I'm still running the calculations, but here we go:

(with a temporary table first:)
```{r}
bincollections <- fread(input = "/stor/azolla_metagenome/Azolla_genus_metagenome/data/curated_bins/summary.bin_info.tab",
                        sep = '\t',
                        fill = TRUE,
                        header = T,
                        stringsAsFactors = T,
                        data.table = T
                        )
```

```{r}
# remove double entries due to double BAT taxonomy
bincollections <- bincollections[classification != 'taxid_assigned_(2/2)' ]

#bincollections$Collection <- 'refined' # fix while waiting for the final table

levels(bincollections$Sample) <- c("_A. filiculoides_<br>'lab'",
                                   "_A. filiculoides_<br>'minus-cyano'",
                                   "_A. filiculoides_<br>'wild'",
                                   "_A. mexicana_",
                                   "_A. microphylla_",
                                   "_A. nilotica_",
                                   "_A. rubra_",
                                   "_A. caroliniana_<br>1",
                                   "_A. caroliniana_<br>2"
                                   )
```


Now I want to rank the bins per sample, per method for completeness. 
I can use this later as an x axis in my plot that is universal for all sample/method combos

```{r}
bincollections[,
               rank := rank(x = -percent_completion,
                            ties.method='first'
                            )
               , by=list(Sample,Collection)]
bincollections[,
               rankCM := rank(x = -Completeness,
                              ties.method='first'
                              )
               , by=list(Sample,Collection)]

# get total nr of bins in small table
bincolcount <- bincollections[,
               .(count = max(rank))
               , by=list(Sample,Collection)]

bincolcountAnvio9010 <- bincollections[(percent_completion > 90 & percent_redundancy < 10),
                        .(count = max(rank))
                        , by=list(Sample,Collection)]

bincolcountCheckM9010 <- bincollections[(Completeness > 90 & Contamination < 10),
                        .(count = max(rank))
                        , by=list(Sample,Collection)]


```


On Anvio stats:
```{r}
attach(bincollections)
# set the basic data and aesthetic
bincol_plot <- ggplot(data = bincollections)
bincol_plot <- bincol_plot + geom_bar(aes(y= percent_completion,
                                          x=rank),
                                      stat='identity',
                                      fill='darkgreen'
                                      )
bincol_plot <- bincol_plot + geom_bar(aes(y=-percent_redundancy,
                                          x=rank),
                                      stat='identity',
                                      fill='red'
                                      )
bincol_plot <- bincol_plot + geom_text(data=bincolcount,
                                       aes(y=80,
                                           x=18,
                                           label=count)
                                       )
bincol_plot <- bincol_plot + geom_abline(mapping = aes(intercept= 90,
                                                       slope=0),
                                         col='grey')
bincol_plot <- bincol_plot + geom_abline(mapping = aes(intercept=-10,
                                                       slope=0),
                                         col='grey')

# facets
bincol_plot <- bincol_plot + facet_grid(Collection ~ Sample)
bincol_plot <- bincol_plot + scale_y_continuous(name = 'Redundancy/Completion'
                                                ,limits = c(-20,100))
bincol_plot <- bincol_plot + scale_x_continuous(name = 'Completeness rank'
                                                ,breaks = c(0,10,20))

# make it pretty
bincol_plot <- bincol_plot + theme_classic()
bincol_plot <- bincol_plot + theme(
                               #axis.text.x = element_blank(),
                               #axis.title.x= element_blank(),
                               legend.position = "bottom",
                               strip.text.x= element_markdown(angle=80,size = 10),
                               strip.text.y= element_markdown(angle=90,size = 10),
                               strip.background = element_blank()
                               )
# and go!
bincol_plot
```

the same with CheckM

```{r}
attach(bincollections)
# set the basic data and aesthetic
bincol_plot <- ggplot(data = bincollections)
bincol_plot <- bincol_plot + geom_bar(aes(y= Completeness,
                                          x=rankCM),
                                      stat='identity',
                                      fill='darkgreen'
                                      )
bincol_plot <- bincol_plot + geom_bar(aes(y=-Contamination,
                                          x=rank),
                                      stat='identity',
                                      fill='red'
                                      )
bincol_plot <- bincol_plot + geom_text(data=bincolcount,
                                       aes(y=80,
                                           x=18,
                                           label=count)
                                       )
bincol_plot <- bincol_plot + geom_abline(mapping = aes(intercept= 90,
                                                       slope=0),
                                         col='grey')
bincol_plot <- bincol_plot + geom_abline(mapping = aes(intercept=-10,
                                                       slope=0),
                                         col='grey')

# facets
bincol_plot <- bincol_plot + facet_grid(Collection ~ Sample)
bincol_plot <- bincol_plot + scale_y_continuous(name = 'Redundancy/Completion'
                                                ,limits = c(-20,100))
bincol_plot <- bincol_plot + scale_x_continuous(name = 'Completeness rank'
                                                ,breaks = c(0,10,20))

# make it pretty
bincol_plot <- bincol_plot + theme_classic()
bincol_plot <- bincol_plot + theme(
                               #axis.text.x = element_blank(),
                               #axis.title.x= element_blank(),
                               legend.position = "bottom",
                               strip.text.x= element_markdown(angle=80,size = 10),
                               strip.text.y= element_markdown(angle=90,size = 10),
                               strip.background = element_blank()
                               )
# and go!
bincol_plot
```


## ideas for a similar thing with taxonomy level 

First, I want to "melt" all the taxonomy levels into two collumns rather than a collumn per level.

```{r}
bincolmelt <- data.table::melt.data.table(data = bincollections
                                          ,id.vars = c('Collection',
                                                       'Sample')
                                          ,measure.vars = c('superkingdom',
                                                            'phylum',
                                                            'class',
                                                            'order',
                                                            'family',
                                                            'genus',
                                                            'species'
                                                            )
                                          ,variable.name = 'taxlevel'
                                          ,value.name    = 'taxvalue'
                                          ,na.rm = F
                                          ,variable.factor = T
                                          ,value.factor = T
                                          )
bincolmelt
```


```{r}
# set the basic data and aesthetic
bincol_plot <- ggplot(data = bincolmelt[taxvalue=='no_support'])
bincol_plot <- bincol_plot + geom_bar(aes(x=Collection,
                                          fill=Collection))


# facets
bincol_plot <- bincol_plot + facet_grid(Sample ~ taxlevel)
bincol_plot <- bincol_plot + scale_fill_brewer(type = 'qual',palette = 2)

# make it pretty
bincol_plot <- bincol_plot + theme_classic()
bincol_plot <- bincol_plot + xlab('Binning method') + ylab('Count of bins without taxonomy assigned')
bincol_plot <- bincol_plot + theme(
                               #axis.text.x = element_markdown(angle=45,hjust = 1),
                               axis.text.x = element_blank(),
                               #axis.title.x= element_blank(),
                               legend.position = "bottom",
                               strip.text.x= element_markdown(angle=60,size = 10),
                               strip.text.y= element_markdown(angle=0,size = 10),
                               strip.background = element_blank()
                               )
# and go!
bincol_plot
```


```{r}
# set the basic data and aesthetic
bincol_plot <- ggplot(data = bincolmelt[taxvalue=='no_support'])
bincol_plot <- bincol_plot + geom_bar(aes(x=Collection,
                                         fill=taxlevel)
                                     ,position = 'dodge')

# facets
bincol_plot <- bincol_plot + facet_grid(Sample ~ .)
bincol_plot <- bincol_plot + scale_fill_brewer(type = 'qual',palette = 2)

# make it pretty
bincol_plot <- bincol_plot + theme_classic()
bincol_plot <- bincol_plot + xlab('Binning method') + ylab('Count of bins without taxonomy assigned')
bincol_plot <- bincol_plot + theme(
                               #axis.text.x = element_markdown(angle=45,hjust = 1),
                               axis.text.x = element_blank(),
                               #axis.title.x= element_blank(),
                               legend.position = "bottom",
                               strip.text.x= element_markdown(angle=60,size = 10),
                               strip.text.y= element_markdown(angle=0,size = 10),
                               strip.background = element_blank()
                               )
# and go!
bincol_plot
```
