---
title: "Statistics on Bins in the Azolla metagenome project"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document: default
---

Import the data:
```{r, knitr root.dir="/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/"}
#setwd("/stor/azolla_metagenome/Azolla_genus_metagenome/analyses/")
library(data.table)
library(ggplot2)
bins <- fread(input = "/stor/azolla_metagenome/Azolla_genus_metagenome/data/curated_bins/refined.bin_info.tab",
              sep = '\t',
              fill = TRUE,
              header = T,
              stringsAsFactors = T,
              data.table = T
)
```


```{r}
azfil_wild_abundance <- fread(input = "../data/curated_bins/refined/Azfil_wild/bins_across_samples/mean_coverage.txt",
                              sep = '\t',
                              header= T,
                              stringsAsFactors = T,
                              data.table = T,
                              fill= T
                              )
names(azfil_wild_abundance) <- sub('assembly_hybrid_doublefiltered_sample_Azfil_wild_binningsignal_',
                                   '',
                                   names(azfil_wild_abundance)
                                   )

Azfil_ratios <- azfil_wild_abundance[,.(bin=bins,
                                        ratio_leaf_over_plant=((Azfil_wild_galgw_E_1+Azfil_wild_galgw_E_2+Azfil_wild_galgw_E_3)/3)
                                                             /((Azfil_wild_galgw_P_2+Azfil_wild_galgw_P_3+Azfil_wild_galgw_P_4)/3),
                                        pacbio_mean_coverage=pacbio_reads)]
Azfil_ratios[,.(bin,
                ratio_leaf_over_plant=round(ratio_leaf_over_plant,2),
                pacbio_mean_coverage=round(pacbio_mean_coverage,2))]
```

Scroll through the whole table here:
```{r}
bins
```

Now, look at a smaller table with all bins, their origin sample, size, completion and redundancy:
```{r}
bins[,
     .(length_mb     = round(sum(total_length/1000000),2),
       anvio_compl   = round(sum(percent_completion),0),
       anvio_redun   = round(sum(percent_redundancy),0),
       checkm_compl  = round(sum(Completeness),0),
       checkm_redun  = round(sum(Contamination),0)
       ),
     by=c('Sample', 'bins')
     ]
```

How many bins do we have per sample
```{r}

bins_per_sample <- bins[,
                        .(bincount=length(bins)),
                        by='Sample'
                        ]
bins_per_sample[order(-bins_per_sample[,2])]
```


Next, pivot sample origin versus taxonomy and either count bins, or display size in mbase

First, count bins in class:
```{r}
dcast.data.table(data = bins[class != NA | class != 'no_support']
                 ,formula = Sample ~ class 
                 ,fun.aggregate = length
                 ,value.var = 'total_length'
                 )
```

Second, count bins in order and show only orders that have more than 5 bins over the entire genus metagenome.
```{r,}
#knitr::kable(
tmpdt <- dcast.data.table(data = bins[order != NA | order != 'no_support']
                          ,formula = order ~ Sample
                          ,fun.aggregate = length
                          ,value.var = 'total_length'
                          )
#data.table::all
#knitr::kable(tmpdt)
threshold <- 5
vector <- (apply(X = tmpdt[,2:length(tmpdt)],MARGIN = 1,FUN = 'sum') >= threshold)
tmpdt[vector]
```

Now select the bins from azfil lab that meet the above threshold (and remove the nostocales for now)
```{r}
#interesting orders:
i_orders <- tmpdt[(vector & order != 'Nostocales')]$order

bins[Sample == 'Azfil_lab' &
     order  %in% i_orders  ]$bins
```

Now let's look at that table with all bins again, but filter for the orders we defined above

```{r}
bins[order %in% i_orders,
     .(length_mb     = round(sum(total_length/1000000),2),
       anvio_compl   = round(sum(percent_completion),0),
       anvio_redun   = round(sum(percent_redundancy),0),
       checkm_compl  = round(sum(Completeness),0),
       checkm_redun  = round(sum(Contamination),0)
       ),
     by=c('Sample', 'bins')
     ]
```

